File: app/src/main/java/com/elvotra/clean/presentation/model/mapper/PostDetailsViewItemMapper.java
package com.elvotra.clean.presentation.model.mapper;

import com.elvotra.clean.domain.model.Comment;
import com.elvotra.clean.domain.model.Post;
import com.elvotra.clean.domain.model.User;
import com.elvotra.clean.presentation.model.PostCommentViewItem;
import com.elvotra.clean.presentation.model.PostDetailsViewItem;
import com.elvotra.clean.utils.Constants;

import java.util.ArrayList;
import java.util.List;

public class PostDetailsViewItemMapper {

    private static PostDetailsViewItemMapper INSTANCE;

    public static PostDetailsViewItemMapper getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new PostDetailsViewItemMapper();
        }
        return INSTANCE;
    }

    private PostDetailsViewItemMapper() {
    }

    public PostDetailsViewItem transform(Post post) {
        List<Comment> comments = post.getComments();
        List<PostCommentViewItem> postCommentViewItems = new ArrayList<>(comments.size());
        for (int i = 0; i < comments.size(); i++) {
            Comment comment = comments.get(i);
            String avatar = Constants.AVATAR_BASE_URL + ((comment.getEmail() != null) ? comment.getEmail() : "");
            PostCommentViewItem commentViewItem = new PostCommentViewItem(comment.getId(), comment.getName(), comment.getEmail(), avatar, comment.getBody());
            postCommentViewItems.add(commentViewItem);
        }
        User user = post.getUser();
        return transform(post, user, postCommentViewItems);
    }

    public PostDetailsViewItem transform(Post post, User user, List<PostCommentViewItem> commentViewItems) {
        String username = (user != null) ? user.getName() : "Unknown";
        String useremail = (user != null && user.getEmail() != null) ? user.getEmail() : "";
        String avatar = Constants.AVATAR_BASE_URL + ((user != null && user.getEmail() != null) ? user.getEmail() : "");

        return new PostDetailsViewItem(post.getId(), username, useremail, avatar, post.getTitle(), post.getBody(), commentViewItems);
    }

}

File: app/src/main/java/com/elvotra/clean/presentation/model/mapper/PostDetailsViewItemMapper.kt
package com.elvotra.clean.presentation.model.mapper

import com.elvotra.clean.domain.model.Comment
import com.elvotra.clean.domain.model.Post
import com.elvotra.clean.domain.model.User
import com.elvotra.clean.presentation.model.PostCommentViewItem
import com.elvotra.clean.presentation.model.PostDetailsViewItem
import com.elvotra.clean.utils.Constants
import java.util.*

class PostDetailsViewItemMapper private constructor() {

    fun transform(post: Post): PostDetailsViewItem {
        val comments = post.comments
        val postCommentViewItems = ArrayList<PostCommentViewItem>()
        comments?.forEach { comment: Comment ->
            val (_, id, name, email, body) = comment
            val avatar = Constants.AVATAR_BASE_URL + email
            val commentViewItem = PostCommentViewItem(id, name, email, avatar, body)
            postCommentViewItems.add(commentViewItem)
        }
        val user = post.user
        return transform(post, user, postCommentViewItems)
    }

    fun transform(post: Post, user: User?, commentViewItems: List<PostCommentViewItem>): PostDetailsViewItem {
        val username = user?.name ?: "Unknown"
        val useremail = if (user != null && user.email != null) user.email else ""
        val avatar = Constants.AVATAR_BASE_URL + if (user != null && user.email != null) user.email else ""

        return PostDetailsViewItem(post.id, username, useremail, avatar, post.title, post.body, commentViewItems)
    }

    companion object {

        private var INSTANCE: PostDetailsViewItemMapper? = null

        fun instance(): PostDetailsViewItemMapper {
            return PostDetailsViewItemMapper.INSTANCE ?: PostDetailsViewItemMapper()
                    .apply { PostDetailsViewItemMapper.INSTANCE = this }
        }
    }
}
